### PR TITLE
N3template operation

### PR DESCRIPTION
**Pull request by VeljkoMaksimovic**
Create N3Template for inserting new triples. Needs to be expanded to support modification and removal of existing triples

### CODE REVIEW

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 1]
```diff
@@ -34,13 +52,89 @@ public Parameters getProvidedParams() {
...
+	private String insertParameters(OperationData input) throws InputMismatchException{
+
+		EditN3GeneratorVTwo gen = new EditN3GeneratorVTwo();
+
+		//Had to convert String to List<String> to be compatible with EditN3GeneratorVTwo methods
+		List<String> n3WithParameters = Arrays.asList(n3Text);
+
+		//region Substitute IRI variables
+		Map<String, List<String>> parametersToUris = requiredParams.params.values().stream()
+				.filter(value->value.getType().getName()=="anyURI")
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:98]: What is "anyURI"? Maybe create a constant for it?

Comment (VeljkoMaksimovic) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:98]: XSD Datatype for Uniform Resource Identifier (http://books.xmlschemata.org/relaxng/ch19-77009.html). Ok, I will create anyURI constant later today.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 2]
```diff
@@ -34,13 +52,89 @@ public Parameters getProvidedParams() {
...
+		String[] leftoverVariables = Arrays.stream(n3WithParameters.get(0).split(" "))
+				.filter(token -> token.startsWith("?")).toArray(String[]::new);
+		if(leftoverVariables.length>0){
+			throw new InputMismatchException("N3 template variables:'" + Arrays.toString(leftoverVariables) +
+					"' dont have corresponding input parameters with values that should be inserted in their place.");
+		}
+		return n3WithParameters.get(0);
 	}
 
+	private boolean isInputValid(OperationData input) {
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:124]: I have transformed Operation from an interface to an abstract class and move there this method. Here might be the git merge conflict.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/types/ParameterType.java [hunk 3]
```diff
@@ -13,7 +13,9 @@
 	public RDFDatatype getRDFDataType() {
 		return new XSDDatatype(name);
 	}
-	
+
+	public String getName() { return this.name; }
```
[the remark this answers is not in the document]
Comment (VeljkoMaksimovic) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/types/ParameterType.java:17]: I have been using it to compare if two Datatypes are equal. I will try to find another way to do this, without using name property of the given datatype

File: home/src/main/resources/rdf/abox/filegraph/dynamic-api-individuals.n3 [hunk 4]
```diff
@@ -64,6 +64,15 @@
         rdfs:label              "Integer type"@en-US ;
         vitro:mostSpecificType  dynapi:primitiveParameterType .
 
+<https://vivoweb.org/ontology/vitro-dynamic-api/parameter/type/XSDanyURI>
```
[the remark this answers is not in the document]
Comment (VeljkoMaksimovic) [reply home/src/main/resources/rdf/abox/filegraph/dynamic-api-individuals.n3:67]: Well I'm not sure to be honest. It seems redundant to create custom validators for URIs, when XSDDatatype has a specific datatype for them. Am I missing some benefits with custom validator approach?

[the remark this answers is not in the document]
Comment (chenejac) [reply home/src/main/resources/rdf/abox/filegraph/dynamic-api-individuals.n3:67]: There is in RDF, but not in JSON, loading of JSON data to Java structure is separated from the validation, in the validation phase Parameter Types might be checked. I will work on validation tomorrow, and inform you if you should change something here.

**Review by chenejac — APPROVED** [review]

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 5]
```diff
@@ -30,21 +50,86 @@ public Parameters getRequiredParams() {
...
+
+	private String insertParameters(OperationData input) throws InputMismatchException{
+
+		EditN3GeneratorVTwo gen = new EditN3GeneratorVTwo();
+
+		//Had to convert String to List<String> to be compatible with EditN3GeneratorVTwo methods
+		List<String> n3WithParameters = Arrays.asList(n3Text);
+
+		//region Substitute IRI variables
+		Map<String, List<String>> parametersToUris = requiredParams.getParameters().values().stream()
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:100]: Wouldn't it be better to have parameters filtering logic in Parameters class?

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 6]
```diff
@@ -30,21 +50,86 @@ public Parameters getRequiredParams() {
...
+				.filter(value->value.getRDFDataType().getURI().equals(ANY_URI))
+				.collect(Collectors.toMap(param -> param.getName(), param -> Arrays.asList(input.get(param.getName()))));
+
+		gen.subInMultiUris(parametersToUris, n3WithParameters);
+		//endregion
+
+		//region Substitute other (literal) variables
+		Map<String, List<Literal>> parametersToLiterals = requiredParams.getParameters().values().stream()
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:108]: This looks very similar to code above. Also might need to consider moving it to Parameters class

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java (hunk 5 above)
Comment (VeljkoMaksimovic) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:100]: Seems like a valid point. I will move both of these methods within Parameters class.

**Review by litvinovg — CHANGES_REQUESTED** [review]
Put parameters filtering logic in Parameters class.

File: home/src/main/resources/rdf/abox/filegraph/dynamic-api-individuals.n3 (hunk 4 above)
[the remark this answers is not in the document]
Comment (VeljkoMaksimovic) [reply home/src/main/resources/rdf/abox/filegraph/dynamic-api-individuals.n3:67]: Is this change still required (Do I need to write a validator for URIs?)

[the remark this answers is not in the document]
Comment (chenejac) [reply home/src/main/resources/rdf/abox/filegraph/dynamic-api-individuals.n3:67]: No, there is anyURI data type implemented in the dynamic api. Therefore, just ignore this comment, I will mark it is as resolved.

**Review by litvinovg — APPROVED** [review]

**Pull request merged**